File: Ejemplos/Ficheros-Java/src/main/java/es/joseluisgs/texto/Texto.java
package es.joseluisgs.texto;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Stream;

public class Texto {
    public static void main(String[] args) {
        leerFichero();
        escribirFichero();
        copiarFichero();
    }

    private static void copiarFichero() {
        String myPath = System.getProperty("user.dir");
        String myFile = myPath + File.separator + "data" + File.separator + "fichero-texto.txt";
        System.out.println(myFile);
        // Existe y podemos leerlo?
        File fOrigen = new File(myFile);
        if (fOrigen.exists() && fOrigen.canRead()) {
            System.out.println("Existe y podemos leerlo");
        } else {
            System.out.println("No existe o no podemos leerlo");
        }
        String myCopy = myPath + File.separator + "data" + File.separator + "fichero-texto-copy.txt";
        File fCopy = new File(myCopy);

        // leemos con BufferReader/FileReader y escribimos con PrintWriter/FileWriter
        try (BufferedReader br = new BufferedReader(new FileReader(fOrigen))) {
            // OJO si no existe lo crea, si existe lo sobreescribe
            try (PrintWriter pw = new PrintWriter(new FileWriter(fCopy))) {
                String linea;
                while ((linea = br.readLine())!= null) {
                    pw.println(linea);
                }
            }
        } catch (IOException e) {
            System.out.println("Error al copiar" + e.getMessage());
        }

        try (BufferedReader br = new BufferedReader(new FileReader(fOrigen))) {
            // OJO si no existe lo crea, si no añade al final
            try (PrintWriter pw = new PrintWriter(new FileWriter(fCopy, true))) {
                String linea;
                while ((linea = br.readLine())!= null) {
                    pw.println(linea);
                }
            }
        } catch (IOException e) {
            System.out.println("Error al copiar" + e.getMessage());
        }

        // Usando NIO y api stream y printWriter
        // Si el fichero es muy grande, lo escribimos en trozos
        try (Stream<String> stream = Files.lines(fOrigen.toPath())) {
            try (PrintWriter pw = new PrintWriter(new FileWriter(fCopy))) {
                stream.forEach(pw::println);
            }
        } catch (IOException e) {
            System.out.println("Error al copiar" + e.getMessage());
        }

        // Si no es muy grande lo leemos todo y lo escribimos linea a linea
        try {
            List<String> lineas = Files.readAllLines(fOrigen.toPath());
            try (PrintWriter pw = new PrintWriter(new FileWriter(fCopy))) {
                for (String linea : lineas) {
                    pw.println(linea);
                }
            }
        } catch (IOException e) {
            System.out.println("Error al copiar" + e.getMessage());
        }

    }

    private static void escribirFichero() {
        String myPath = System.getProperty("user.dir");
        String myFile = myPath + File.separator + "data" + File.separator + "fichero-texto-2.txt";
        System.out.println(myFile);

        File f = new File(myFile);
        if (f.exists()) {
            System.out.println("Existe");
        } else {
            System.out.println("No existe");
        }
        System.out.println();

        // Escribimos fichero
        try {
            BufferedWriter bw = new BufferedWriter(new FileWriter(f));
            bw.write("Hola");
            bw.newLine();
            bw.write("Adios");
            bw.newLine();
            bw.close();
        } catch (IOException e) {
            System.out.println("Error de escritura");
        } finally {
            System.out.println("Fin de escritura");
        }

        // autoClose try catch
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(f))) {
            bw.write("Hola");
            bw.newLine();
            bw.write("Adios");
            bw.newLine();
        } catch (IOException e) {
            System.out.println("Error de escritura");
        } finally {
            System.out.println("Fin de escritura");
        }

        // PrintWriter
        try {
            PrintWriter pw = new PrintWriter(new FileWriter(f));
            pw.println("Hola");
            pw.println("Adios");
            pw.close();
        } catch (IOException e) {
            System.out.println("Error de escritura");
        } finally {
            System.out.println("Fin de escritura");
        }

        // PrintWriter autoClose
        try (PrintWriter pw = new PrintWriter(new FileWriter(f))) {
            pw.println("Hola");
            pw.println("Adios");
        } catch (IOException e) {
            System.out.println("Error de escritura");
        } finally {
            System.out.println("Fin de escritura");
        }


    }

    private static void leerFichero() {
        System.out.println("Hola Leer Ficheros");

        // Directorio de ejecución
        System.out.println(System.getProperty("user.dir"));
        String myPath = System.getProperty("user.dir");
        System.out.println(myPath);
        String myFile = myPath + File.separator + "data" + File.separator + "fichero-texto.txt";
        System.out.println(myFile);

        File f = new File(myFile);
        if (f.exists()) {
            System.out.println("Existe");
        } else {
            System.out.println("No existe");
        }
        System.out.println();

        // Existe??
        if (f.exists()) {
            System.out.println("Existe");
        } else {
            System.out.println("No existe");
        }
        System.out.println();

        // Es un directorio?
        if (f.isDirectory()) {
            System.out.println("Es un directorio");
        } else {
            System.out.println("No es un directorio");
        }
        System.out.println();

        // Tenemos permisos de lectura?
        if (f.canRead()) {
            System.out.println("Tenemos permisos de lectura");
        } else {
            System.out.println("No tenemos permisos de lectura");
        }

        System.out.println();
        BufferedReader brFuera = null;
        try {
            brFuera = new BufferedReader(new FileReader(f));
            // Vamos leyendo linea a linea hasta que no haya más
            String linea = brFuera.readLine();
            while (linea != null) {
                // Procesamos la línea
                System.out.println(linea);
                linea = brFuera.readLine();
            }
        } catch (IOException e) {
            System.out.println("Error de lectura");
        } finally {
            System.out.println("Fin de lectura");
            if (brFuera != null) {
                try {
                    brFuera.close();
                } catch (IOException e) {
                    System.out.println("Error al cerrar el fichero");
                }
            }
        }

        // Leemos con bufferReader
        try {
            BufferedReader br = new BufferedReader(new FileReader(f));
            // Vamos leyendo linea a linea hasta que no haya más
            String linea = br.readLine();
            while (linea != null) {
                // Procesamos la línea
                System.out.println(linea);
                linea = br.readLine();
            }
            br.close();
        } catch (IOException e) {
            System.out.println("Error de lectura");
        } finally {
            System.out.println("Fin de lectura");
        }
        System.out.println();

        // Con un inputStream
        try {
            InputStreamReader is = new InputStreamReader(new FileInputStream(f));
            BufferedReader br = new BufferedReader(is);
            // Vamos leyendo linea a linea hasta que no haya más
            String linea = br.readLine();
            while (linea != null) {
                // Procesamos la línea
                System.out.println(linea);
                linea = br.readLine();
            }brFuera.close();
            br.close();
        }
        catch (IOException e) {
            System.out.println("Error de lectura");
        }
        finally {
            System.out.println("Fin de lectura");
        }

        // Con autoclese try catch
        try (BufferedReader br = new BufferedReader(new FileReader(f))) {
            // Vamos leyendo linea a linea hasta que no haya más
            String line;
            while((line = br.readLine()) != null) {
                // Procesamos la línea
                System.out.println(line);
            }
        } catch (IOException e) {
            System.out.println("Error de lectura");
        }

        // Con autoclese try catch y Scanner
        try (Scanner sc = new Scanner(f)) {
            // Vamos leyendo linea a linea hasta que no haya más
            while(sc.hasNextLine()) {
                // Procesamos la línea
                System.out.println(sc.nextLine());
            }
        } catch (IOException e) {
            System.out.println("Error de lectura");
        }
        System.out.println();

        // Con la api stream
        // // Ideal cuando tenemos ficheros muy grandes y procesamos linea a linea
        try (Stream<String> stream = Files.lines(f.toPath())) {
            stream.forEach(System.out::println); // Procesamos linea a linea
        } catch (IOException e) {
            System.out.println("Error de lectura");
        }

        try {
            // Procesamos linea a linea
            Files.readAllLines(f.toPath()).forEach(System.out::println);
        } catch (IOException e) {
            System.out.println("Error de lectura");
        }
    }
}

File: Ejemplos/Ficheros-Kotlin/src/main/kotlin/texto/Texto.kt
package texto

import java.io.File

fun main() {
    println("Hola ficheros de texto!")
    // leerFicheroTexto()
    // escribirFicheroTexto()
    copiarFicheroTexto()
}

fun copiarFicheroTexto() {
    val programPath = System.getProperty("user.dir")
    val fileOrigenPath = "$programPath${File.separator}data${File.separator}fichero-texto.txt"
    val fileDestinoPath = "$programPath${File.separator}data${File.separator}fichero-texto-copy.txt"

    val fileOrigen = File(fileOrigenPath)
    val fileDestino = File(fileDestinoPath)

    if (fileOrigen.exists() && fileOrigen.canRead()) {
        println("Copiando fichero...")
    } else {
        println("No se puede copiar el fichero")
        return
    }

    // Copia el fichero
    // Leemos todas las líneas y las escribimos en el fichero destino
    fileDestino.writeText("") // Limpiamos el fichero
    fileOrigen.readLines().forEach { fileDestino.appendText("$it\n") }

    fileDestino.writeText("") // Limpiamos el fichero
    // ForEachLine es equivalente a readLines().forEach
    fileOrigen.forEachLine { fileDestino.appendText("$it\n") }

    // Cuidado con ficheros grandes
    fileOrigen.readText().also { fileDestino.writeText(it) }

    // forEachLine y printWriter
    fileOrigen.forEachLine { fileDestino.printWriter().println(it) }

    // forEachLine y bufferedWriter
    fileOrigen.forEachLine { fileDestino.bufferedWriter().write("$it\n") }


    // Usando buffer de lectura o escritura
    fileOrigen.bufferedReader().useLines { lines ->
        fileDestino.bufferedWriter().use { out ->
            lines.forEach { out.write("$it\n") }
        }
    }

    // con todas las combinaciones que conozcas


}

fun escribirFicheroTexto() {
    val programPath = System.getProperty("user.dir")
    val filePath = "$programPath${File.separator}data${File.separator}fichero-texto-2.txt"
    println("Leyendo fichero: $filePath")

    val fileDestino = File(filePath)

    // Con writeText escribimos todo el contenido del fichero
    fileDestino.writeText("Hola mundo")

    // Podemos añadir contenido al fichero
    fileDestino.appendText("Hola de nuevo")

    // Podemos usar el método writeBytes para escribir bytes
    fileDestino.writeBytes("Hola mundo".toByteArray())

    // printWriter
    fileDestino.printWriter().use { it.println("Hola mundo") }

    // bufferedWriter
    fileDestino.bufferedWriter().use { it.write("Hola mundo") }

    // bufferedWriter + append
    fileDestino.bufferedWriter().use { it.append("Hola mundo") }

    // printWriter + append
    fileDestino.printWriter().use { it.append("Hola mundo") }

}

fun leerFicheroTexto() {
    val programPath = System.getProperty("user.dir")
    val filePath = "$programPath${File.separator}data${File.separator}fichero-texto.txt"
    println("Leyendo fichero: $filePath")

    val fileOrigen = File(filePath)

    // existe
    if (fileOrigen.exists()) {
        println("El fichero existe")
    } else {
        println("El fichero no existe")
    }

    // Si puedo leerlo
    if (fileOrigen.canRead()) {
        println("El fichero se puede leer")
    } else {
        println("El fichero no se puede leer")
    }
    println()

    // Devuelve una lista de líneas
    fileOrigen.readLines().forEach { println(it) }
    println()

    // Procesa linea a linea
    fileOrigen.forEachLine { println(it) }

    // Lee el fichero completo en una cadena
    fileOrigen.readText().also { println(it) }

    // Usa usa una secuencia y se le procesa un callback en el lambda
    fileOrigen.useLines { it.toList() }.forEach { println(it) }
    // fileOrigen.useLines { it.forEach { line -> println(line) } }
    println()

    // Buffer de lectura
    // Lee el fichero completo en una lista de cadena
    fileOrigen.bufferedReader().readLines().forEach { println(it) }
    fileOrigen.bufferedReader().forEachLine { println(it) }
    println()
    fileOrigen.bufferedReader().readLine().let { println(it) }
    println()
    // Lee todo el fichero en una cadena
    fileOrigen.bufferedReader().readText().let { println(it) }
    println()

    // Use equivale al try-with-resources de Java
    // Dentro del use podemos hacer uso de los métodos de BufferedReader
    fileOrigen.bufferedReader().use { it.readText() }.let { println(it) }
    println()

    // InputStream
    fileOrigen.inputStream().readBytes().toString(Charsets.UTF_8).let { println(it) }
    println()

    // InputStreamReader
    fileOrigen.inputStream().reader().readText().let { println(it) }
    println()

    // InputStream + BufferedReader
    fileOrigen.inputStream().bufferedReader().readText().let { println(it) }
    println()

    // InputStreamReader + BufferedReader + use
    fileOrigen.inputStream().reader().buffered().use { it.readText() }.let { println(it) }
    println()
}
